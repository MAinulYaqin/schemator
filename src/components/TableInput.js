import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { MdClose } from "react-icons/lib/md";

import "typeface-roboto";

const Container = styled.div`
  width: 100%;
  height: 36px;
  display: flex;
  align-items: center;
  justify-content: stretch;
  padding: 0px 4px;
  box-sizing: border-box;
`;

export const Select = styled.select`
  background: #333;
  color: white;
  font-family: Roboto;
  font-size: 11px;
  outline: none;
  width: 100%;
  height: 24px;
`;

export const Input = styled.input`
  width: 100%;
  height: 24px;
  color: white;
  background: #111;
  border: 0px;
  outline: none;
  font-family: Roboto;
  font-size: 12px;
  box-sizing: border-box;
  padding: 0px 8px;
`;

export const CloseButton = styled(MdClose)`
  color: #444;
  font-size: 16px;
  padding: 8px;
  cursor: pointer;
`;

const Column = styled.div`
  flex: 1;
  padding: 4px;
`;

const TableInput = ({
  name,
  type,
  onClickRemove,
  onChangeName,
  onChangeType
}) => {
  const types = [
    "BINARY",
    "BOOLEAN",
    "CHAR",
    "DATE",
    "DATE_TIME",
    "DATE_TIME_TZ",
    "DECIMAL",
    "DOUBLE",
    "ENUM",
    "FLOAT",
    "INTEGER",
    "IP_ADDRESS",
    "JSON",
    "JSONB",
    "MAC_ADDRESS",
    "MORPHS",
    "NULLABLE_MORPHS",
    "STRING",
    "TEXT",
    "TIME",
    "TIME_TZ",
    "TIMESTAMP",
    "TIMESTAMP_TZ",
    "UUID"
  ];

  return (
    <Container>
      <Column>
        <Input type="text" value={name} onChange={onChangeName} />
      </Column>
      <Column>
        <Select value={type} onChange={onChangeType}>
          {types.map((type, index) => (
            <option key={index} value={type}>
              {type}
            </option>
          ))}
        </Select>
      </Column>
      <CloseButton onClick={onClickRemove} />
    </Container>
  );
};

TableInput.propTypes = {
  name: PropTypes.string,
  type: PropTypes.string,
  onClickRemove: PropTypes.func,
  onChangeName: PropTypes.func,
  onChangeType: PropTypes.func
};

export default TableInput;
