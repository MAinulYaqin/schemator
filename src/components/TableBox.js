import React, { forwardRef } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import "typeface-roboto";

import TableHeader from "./TableHeader";
import TableButton from "./TableButton";
import TableInput from "./TableInput";
import TableOption from "./TableOption";

const Container = styled.div`
  width: 240px;
  background: #222;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.5);
  box-sizing: border-box;
  border: 2px solid #111;
`;

const TableBox = forwardRef(
  (
    {
      position,
      name,
      fields,
      options,
      onChangeName,
      onChangeFieldName,
      onChangeFieldType,
      onChangeOptions,
      onClickAddField,
      onClickRemoveField,
      onMouseDown,
      onMouseMove,
      onMouseEnter,
      onMouseLeave
    },
    ref
  ) => {
    const { x, y } = position;
    const headerHeight = 36;
    const buttonHeight = 36;
    const fieldHeight = 36 * fields.length;
    const optionHeight = 72;
    const height = headerHeight + buttonHeight + fieldHeight + optionHeight + 4;

    return (
      <foreignObject
        ref={ref}
        x={x}
        y={y}
        width={240}
        height={height}
        onMouseDown={onMouseDown}
        onMouseMove={onMouseMove}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
      >
        <Container>
          <TableHeader caption={name} onChangeCaption={onChangeName} />
          {fields.map(field => (
            <TableInput
              key={field.id}
              name={field.name}
              type={field.type}
              onChangeName={event => onChangeFieldName(event, field.id)}
              onChangeType={event => onChangeFieldType(event, field.id)}
              onClickRemove={() => onClickRemoveField(field.id)}
            />
          ))}
          <TableOption onChange={onChangeOptions} value={options} />
          <TableButton caption="Add New Field" onClick={onClickAddField} />
        </Container>
      </foreignObject>
    );
  }
);

TableBox.propTypes = {
  position: PropTypes.object,
  name: PropTypes.string,
  fields: PropTypes.array,
  options: PropTypes.object,
  onMouseDown: PropTypes.func,
  onMouseMove: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  onClickAddField: PropTypes.func,
  onClickRemoveField: PropTypes.func,
  onChangeFieldName: PropTypes.func,
  onChangeFieldType: PropTypes.func,
  onChangeName: PropTypes.func,
  onChangeOptions: PropTypes.func
};

export default TableBox;
