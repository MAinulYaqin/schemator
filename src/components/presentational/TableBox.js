import React, { forwardRef } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import "typeface-roboto";

import TableHeader from "./TableHeader";
import TableButton from "./TableButton";
import TableInput from "./TableInput";
import TableOption from "./TableOption";

const Container = styled.div`
  width: 240px;
  background: #222;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.5);
  box-sizing: border-box;
  border: 2px solid #111;
`;

const TableBox = forwardRef(
  (
    {
      position,
      types,
      name,
      fields,
      options,
      onChangeName,
      onChangeFieldName,
      onChangeFieldType,
      onChangeOptions,
      onClickAddField,
      onClickRemoveField,
      onMouseDown,
      onMouseUp,
      onMouseMove,
      onMouseEnter,
      onMouseLeave,
      onContextMenu
    },
    ref
  ) => {
    const { x, y } = position;
    const headerHeight = 36;
    const buttonHeight = 36;
    const fieldHeight = 36 * fields.length;
    const optionHeight = 72;
    const height = headerHeight + buttonHeight + fieldHeight + optionHeight + 4;

    // Count total needed of TableOption
    const totalOptionChunks = Math.round(options.length / 2);

    // Split options into several chunks.
    // Each chunk contains 2 options.
    const optionChunks = [...Array(totalOptionChunks)].map((_, index) =>
      options.slice(index * 2, (index + 1) * 2)
    );

    return (
      <foreignObject
        ref={ref}
        x={x}
        y={y}
        width={240}
        height={height}
        onMouseDown={onMouseDown}
        onMouseUp={onMouseUp}
        onMouseMove={onMouseMove}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        onContextMenu={onContextMenu}
      >
        <Container>
          <TableHeader caption={name} onChangeCaption={onChangeName} />
          {fields.map(field => (
            <TableInput
              key={field.id}
              types={types}
              name={field.name}
              type={field.type}
              onChangeName={event => onChangeFieldName(event, field.id)}
              onChangeType={event => onChangeFieldType(event, field.id)}
              onClickRemove={() => onClickRemoveField(field.id)}
            />
          ))}
          {optionChunks.map((chunk, index) => (
            <TableOption key={index} onChange={onChangeOptions} items={chunk} />
          ))}
          <TableButton caption="Add New Field" onClick={onClickAddField} />
        </Container>
      </foreignObject>
    );
  }
);

TableBox.propTypes = {
  position: PropTypes.object,
  name: PropTypes.string,
  fields: PropTypes.array,
  options: PropTypes.array,
  onMouseDown: PropTypes.func,
  onMouseUp: PropTypes.func,
  onMouseMove: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  onContextMenu: PropTypes.func,
  onClickAddField: PropTypes.func,
  onClickRemoveField: PropTypes.func,
  onChangeFieldName: PropTypes.func,
  onChangeFieldType: PropTypes.func,
  onChangeName: PropTypes.func,
  onChangeOptions: PropTypes.func
};

export default TableBox;
