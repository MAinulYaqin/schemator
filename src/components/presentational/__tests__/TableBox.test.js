import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import TableInput from "../TableInput";
import TableBox from "../TableBox";

jest.mock("../TableHeader", () => "TableHeader");
jest.mock("../TableInput", () => "TableInput");
jest.mock("../TableOption", () => "TableOption");
jest.mock("../TableButton", () => "TableButton");

describe("TableBox", () => {
  const setup = propOverrides => {
    const props = {
      position: { x: 32, y: 32 },
      types: ["NUMBER", "STRING", "BOOLEAN"],
      name: "User",
      fields: [
        {
          id: "1",
          name: "fullname",
          type: "STRING"
        },
        {
          id: "2",
          name: "username",
          type: "STRING"
        }
      ],
      options: [
        {
          id: "id",
          label: "ID",
          isChecked: true
        },
        {
          id: "rememberToken",
          label: "Remember Token",
          isChecked: false
        }
      ],
      onClickAddField: jest.fn(),
      onClickRemoveField: jest.fn(),
      onChangeFieldName: jest.fn(),
      onChangeFieldType: jest.fn(),
      onChangeName: jest.fn(),
      onChangeOptions: jest.fn(),
      onMouseDown: jest.fn(),
      onMouseUp: jest.fn(),
      onMouseMove: jest.fn(),
      onMouseEnter: jest.fn(),
      onMouseLeave: jest.fn(),
      onContextMenu: jest.fn(),
      ...propOverrides
    };

    const wrapper = shallow(<TableBox {...props} />);

    return {
      wrapper,
      props
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should calls 'onMouseDown' while mouse down", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("mouseDown");

    expect(props.onMouseDown).toBeCalled();
  });

  it("should calls 'onMouseUp' while mouse up", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("mouseUp");

    expect(props.onMouseUp).toBeCalled();
  });

  it("should calls 'onMouseMove' while mouse move", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("mouseMove");

    expect(props.onMouseMove).toBeCalled();
  });

  it("should calls 'onMouseEnter' while mouse enter", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("mouseEnter");

    expect(props.onMouseEnter).toBeCalled();
  });

  it("should calls 'onMouseLeave' while mouse leave", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("mouseLeave");

    expect(props.onMouseLeave).toBeCalled();
  });

  it("should calls 'onContextMenu' while mouse right button clicked", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("contextmenu");

    expect(props.onContextMenu).toBeCalled();
  });

  it("should calls 'onChangeName' while TableInput name changed", () => {
    const { wrapper, props } = setup();
    const fieldID = "1";
    const event = {
      target: {
        value: "name"
      }
    };

    wrapper
      .find(TableInput)
      .at(0)
      .props()
      .onChangeName(event);

    expect(props.onChangeFieldName).toBeCalledWith(event, fieldID);
  });

  it("should calls 'onChangeType' while TableInput type changed", () => {
    const { wrapper, props } = setup();
    const fieldID = "1";
    const event = {
      target: {
        value: "INT"
      }
    };

    wrapper
      .find(TableInput)
      .at(0)
      .props()
      .onChangeType(event);

    expect(props.onChangeFieldType).toBeCalledWith(event, fieldID);
  });

  it("should calls 'onClickRemove' while TableInput remove button clicked", () => {
    const { wrapper, props } = setup();
    const fieldID = "1";

    wrapper
      .find(TableInput)
      .at(0)
      .props()
      .onClickRemove();

    expect(props.onClickRemoveField).toBeCalledWith(fieldID);
  });
});
