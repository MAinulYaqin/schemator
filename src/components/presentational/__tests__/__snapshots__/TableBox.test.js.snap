// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableBox should renders default 1`] = `
<foreignObject
  height={220}
  onContextMenu={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
  width={240}
  x={32}
  y={32}
>
  <styled.div>
    <TableHeader
      caption="User"
      onChangeCaption={[Function]}
    />
    <TableInput
      key="1"
      name="fullname"
      onChangeName={[Function]}
      onChangeType={[Function]}
      onClickRemove={[Function]}
      type="STRING"
    />
    <TableInput
      key="2"
      name="username"
      onChangeName={[Function]}
      onChangeType={[Function]}
      onClickRemove={[Function]}
      type="VARCHAR"
    />
    <TableOption
      items={Array []}
      onChange={[Function]}
    />
    <TableButton
      caption="Add New Field"
      onClick={[Function]}
    />
  </styled.div>
</foreignObject>
`;
